pkg/kp/consulutil/watch.go

```diff
@@ -0,0 +1,165 @@
+package consulutil
+
+import (
+	"time"
+
+	"github.com/square/p2/Godeps/_workspace/src/github.com/hashicorp/consul/api"
+)
+
+// WatchPrefix watches a Consul prefix for changes to any keys that have the prefix. When
+// anything changes, all Key/Value pairs having that prefix will be written to a provided
+// channel.
+//
+// Errors will sent on the given output channel but do not otherwise affect execution. The
+// provided output stream will become owned by this function call, and this call will
+// close it when the function ends. This function will run until explicitly canceled by
+// closing the "done" channel. Data is written to the output channel synchronously, so
+// readers must consume the data to this method will block.
+func WatchPrefix(
+	prefix string,
+	clientKV ConsulLister,
+	outPairs chan<- api.KVPairs,
+	done <-chan struct{},
+	outErrors chan<- error,
+) {
+	defer close(outPairs)
+	var currentIndex uint64
+	timer := time.NewTimer(time.Duration(0))
+
+	for {
+		select {
+		case <-done:
+			return
+		case <-timer.C:
+		}
+		timer.Reset(250 * time.Millisecond) // upper bound on request rate
+		pairs, queryMeta, err := SafeList(clientKV, done, prefix, &api.QueryOptions{
+			WaitIndex: currentIndex,
+		})
+		switch err {
+		case CanceledError:
+			return
+		case nil:
+			currentIndex = queryMeta.LastIndex
+			select {
+			case <-done:
+			case outPairs <- pairs:
+			}
+		default:
+			select {
+			case <-done:
+			case outErrors <- err:
```

Comment:
I think genericized watch is a great opportunity to clean up our error handling and error channels have often been the wrong way to do it, so I'd like to see something different here. I have two (mutually exclusive) suggestions:

1. couple errors and meaningful results together. Instead of having a channel for kvpairs and a channel for errors, have a single channel that outputs a (kvpairs, error) struct. That way callers cannot possibly leave the errors unconsumed - by definition they would have to consume errors and results together. In addition this couples the error to the kvpairs that it was generated on (although probably kvpairs would be nil so that's of questionable value).
2. do not propagate errors out at all and just log them.

I've been reading a lot of rust documentation lately so things like haskell either and rust result are really on my mind. I think if we still want to represent errors via channels, a tuple for both the result and the error is the best approach. (Sadly go doesn't have good sum types but a tuple is still better than separate channels imo)